### PR TITLE
[TESTERS NEEDED] vk: Add remaining pieces

### PR DESCRIPTION
**Pull request by kd-11**
Adds remaining pieces for separate upload path vs 3D rendering. Disabled by default as it still needs to be tuned.
Since passthrough DMA is fixed on NVIDIA when not using separate paths, this technically solves the NVIDIA DMA regressions but probably costs some performance.

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/VKAsyncScheduler.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		void flush(VkSemaphore wait_semaphore = VK_NULL_HANDLE, VkPipelineStageFlags wait_dst_stage_mask = VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT);
+
+		// Thread entry-point
+		void operator()();
+
+		static constexpr auto thread_name = "Vulkan Async Scheduler"sv;
+	};
+
+	using async_scheduler_thread = named_thread<AsyncTaskScheduler>;
+}
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKAsyncScheduler.h:64]: newline

File: rpcs3/Emu/RSX/VK/vkutils/sync.cpp [hunk 2]
```diff
@@ -116,6 +116,24 @@ namespace vk
...
+
+	void event::reset() const
+	{
+		if (m_vk_event) [[likely]]
+		{
+			vkResetEvent(m_device, m_vk_event);
+		}
+		else
+		{
+			*m_value = 0xCAFEBABE;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/sync.cpp:141]: no nullptr check?

File: rpcs3/Emu/RSX/VK/VKAsyncScheduler.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		}
+	}
+
+	void AsyncTaskScheduler::insert_sync_event()
+	{
+		ensure(m_current_cb);
+
+		xqueue_event* sync_label;
+		ensure(m_next_event_id < events_pool_size);
+		sync_label = &m_events_pool[m_next_event_id];
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKAsyncScheduler.cpp:61]: ensure(m_next_event_id < events_pool_size);
		xqueue_event* sync_label = &m_events_pool[m_next_event_id];

File: rpcs3/Emu/RSX/VK/VKAsyncScheduler.cpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		m_sync_required = true;
+
+		// 0. Anything still active?
+		if (m_current_cb)
+		{
+			return m_current_cb;
+		}
+
+		// 1. Check if there is a 'next' entry
+		auto pdev = get_current_renderer();
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKAsyncScheduler.cpp:109]: whats this for?

File: rpcs3/Emu/RSX/VK/vkutils/sync.cpp (hunk 2 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/vkutils/sync.cpp:141]: No. It cannot be null, though an ensure can be added if needed.
This is device memory, if we couldn't allocate it we would have had a vulkan assert.

**Comment by kd-11** [comment]
> TRANSFER
> all nvidia gpu have 2 transfer queque?
> can both be used?

1. No, they only have one proper one with compute support
2. Nope, code uses one at a time

**Comment by kd-11** [comment]
Please do not use this branch unless you're a tester on discord. It will be moved out of draft when it no longer has the capacity to destroy your system. This is very much still in development.

**Comment by kd-11** [comment]
Ok, this branch is now a lot more stable. I have tested rpcs3 running for over a half hour without anything crazy happening so its good enough for testing now.

**Comment by Kravickas** [comment]
NHL 16 in the menu selecting teams

```
·F 0:01:26.258359 {RSX [0x02e3a74]} SIG: Thread terminated due to fatal error: Attempted to transition to an invalid layout
(in file D:\a\1\s\rpcs3\Emu\RSX\VK\vkutils\image_helpers.cpp:126[:24], in function change_image_layout)
```

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/6112045/RPCS3.zip)

**Comment by Satan86** [comment]
Fixes #9752

**Comment by kd-11** [comment]
GPU hangs are fixed as are invalid layout transitions. I think the only thing left is to properly add the options to the GUI

**Comment by Kravickas** [comment]
Nope

> ·F 0:01:14.418846 {RSX [0x02dc0d4]} SIG: Thread terminated due to fatal error: Attempted to transition to an invalid layout
> (in file D:\a\1\s\rpcs3\Emu\RSX\VK\vkutils\image_helpers.cpp:126[:24], in function change_image_layout)
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/6123448/RPCS3.zip)

**Comment by kd-11** [comment]
@Kravickas Where did you get a build for this?

**Comment by Kravickas** [comment]
yikes, looked again and i downloaded the same build 🤦‍♂️

**Comment by Shoegzer** [comment]
@Satan86 Curious how you observed #9752 as fixed?  Still getting black flickering in motorstorm after compiling a build with this PR.

Test system is Linux Mint 20.1 with nvidia gtx 1070 / driver 460.39.  Tested with vulkan.

**Comment by Satan86** [comment]
Retested, it's fixed if Async Texture Streaming is disabled. With it enabled, flickering happens

**Comment by Shoegzer** [comment]
Ah, got it!  Thanks.

**Comment by Shoegzer** [comment]
Thanks.  I noticed there's still minor black screen flickering even with the ATS option off, which I assume is caused by texture upload path bugs as you say.  In any event it's much less pronounced now.

**Comment by Shuneferus** [comment]
Phantom Breaker Extra [BLJM61098] 
Hello Kd-11 this game presents regression with the option enabled: "Asynchronous Texture Streaming" follows captures, log and rsx capture.
The game works very well without the option enabled, I only informed you to help you in the implementation of the new functionality. A hug and thanks for the excellent emulator. 
rpcs3_2021_03_13_16_10_25_147
rpcs3_2021_03_13_16_10_40_977
rpcs3_2021_03_13_16_10_43_043
rpcs3_2021_03_13_16_10_45_326
rpcs3_2021_03_13_16_10_54_541
rpcs3_2021_03_13_16_10_55_308
rpcs3_2021_03_13_16_10_56_924
rpcs3_2021_03_13_16_11_02_590

[BLJM61098_20210313162354_capture.zip](https://github.com/RPCS3/rpcs3/files/6135413/BLJM61098_20210313162354_capture.zip)

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/6135392/RPCS3.log)
System: Windows 10 x64 GPU GTX1060 6GB CPU Intel E5-2660 @3.3Ghz

**Pull request merged**